### PR TITLE
fix: use RFC3896 percent encoding with delta protocol correctness

### PR DESCRIPTION
**Pull request by ion-elgreco**
# Description
Before when I addressed partition encoding I was using the URL encoding, where I naively assumed this was URI encoding compatible, but protocol states to use RFC3896 with additional stricter rules. I have used the percent-encoding-rfc3896 crate for this which forks the percent-encoding crate but adds some extra things to make it it RFC3896 compatible.

The ASCII set contains all reserved characters that need to be encoded according to the protocol

- fixes: #3577

@kdn36

### CODE REVIEW

File: python/tests/test_partition_encoding_rfc3896.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    )
+    print(f"df:\n{df}")
+
+    # write:
+    df.write_delta(tmp_path, delta_write_options={"partition_by": "strings"})
+
+    # read:
+    partitioned_tbl = DeltaTable(tmp_path)
+    pl_df_partitioned = pl.read_delta(partitioned_tbl).sort("a")
+    print(f"pl_df_partitioned:\n{pl_df_partitioned}")
```
Comment (roeap) [inline python/tests/test_partition_encoding_rfc3896.py]: still needed?

File: python/tests/test_partition_encoding_rfc3896.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+from deltalake import DeltaTable
+
+
+@pytest.mark.polars
+def test_partition_encoding_rfc3896(tmp_path):
+    import polars as pl
+    from polars.testing import assert_frame_equal
+
+    (start, stop) = (32, 127)
```
Comment (roeap) [inline python/tests/test_partition_encoding_rfc3896.py]: does this range correspond to the characters we need to encode?

File: crates/core/src/kernel/scalars.rs [hunk 3]
```diff
@@ -86,7 +122,7 @@ impl ScalarExt for Scalar {
         if self.is_null() {
             return NULL_PARTITION_VALUE_DATA_PATH.to_string();
         }
-        encode(Path::from(self.serialize()).as_ref()).to_string()
+        encode_partition_value(self.serialize().as_str()).to_string()
```
Comment (roeap) [inline crates/core/src/kernel/scalars.rs]: hmm, doesn't `encode_partition_value` already return a `String`?

File: crates/core/src/kernel/scalars.rs [hunk 4]
```diff
@@ -9,10 +9,46 @@ use delta_kernel::{
...
 };
-use object_store::path::Path;
+
+use percent_encoding_rfc3986::{utf8_percent_encode, AsciiSet, CONTROLS};
 #[cfg(any(test, feature = "integration_test"))]
 use serde_json::Value;
-use urlencoding::encode;
+
+// ASCII set that needs to be encoded
+const RFC3986_PART: &AsciiSet = &CONTROLS
```
Comment (roeap) [inline crates/core/src/kernel/scalars.rs:18]: nit: would it be more legible to put all characters in an array and then do something like `fold(...)`?

File: crates/core/src/kernel/scalars.rs [hunk 5]
```diff
@@ -9,10 +9,46 @@ use delta_kernel::{
     expressions::{Scalar, StructData},
     schema::StructField,
 };
-use object_store::path::Path;
+
```
Comment (roeap) [inline crates/core/src/kernel/scalars.rs]: should there be a blank line in this import block?

**Review by roeap — COMMENTED** [review]
Just some minor comments.

Would be great to add a link somewhere to the protocol for future reference.

File: python/tests/test_partition_encoding_rfc3896.py (hunk 2 above)
Comment (ion-elgreco) [reply python/tests/test_partition_encoding_rfc3896.py]: It was only the printable characters of ASCII, but I've extended the test range with the extendable ASCII characters

File: python/tests/test_partition_encoding_rfc3896.py (hunk 1 above)
Comment (ion-elgreco) [reply python/tests/test_partition_encoding_rfc3896.py]: Nope :) was just copy pasta from the issue

File: crates/core/src/kernel/scalars.rs (hunk 4 above)
Comment (ion-elgreco) [reply crates/core/src/kernel/scalars.rs:18]: I tried, but const or lazylock doesn't like folding over a slice of bytes, maybe you can get some working code for this?

Comment (ion-elgreco) [reply crates/core/src/kernel/scalars.rs:18]: @roeap shall we merge it now, we can always optimize later, I want to land this before we merge the log replay stuff

**Review by roeap — APPROVED** [review]
👍

File: crates/core/src/kernel/scalars.rs (hunk 4 above)
Comment (roeap) [reply crates/core/src/kernel/scalars.rs:18]: @ion-elgreco - yes :).

**Comment by kdn36** [comment]
Great to see this, thanks a lot. I will test it into polars soon.

One clarifying question on an edge case. Given that `Path` removes leading and trailing slashes, will the encoding be such that any hive partition strings containing a trailing or leading `/` encode correctly and they will not be removed?

**Comment by ion-elgreco** [comment]
@kdn36 trailing slashes will also be encoded yes, you should use the same ASCII set as we use to decode the partitions, since delta protocol partition serialization is more strict than RFC3896

**Comment by kdn36** [comment]
I can confirm that the local tests on polars now pass (using release 1.1.4).

Thanks a lot for your support.
